native/common/include/jp_context.h

```diff
@@ -0,0 +1,310 @@
+	{
+		return m_IsShutdown;
+	}
+
+	/** Release a global reference checking for shutdown.
+	 *
+	 * This should be used in any calls to release resources from a destructor.
+	 * It cannot fail even if the JVM is no longer operating.
+	 */
+	void ReleaseGlobalRef(jobject obj);
+
+	// JPype services
+
+	JPProxyFactory* getProxyFactory()
+	{
+		return m_ProxyFactory;
+	}
+
+	JPTypeManager* getTypeManager()
+	{
+		return m_TypeManager;
+	}
+
+	JPClassLoader* getClassLoader()
+	{
+		return m_ClassLoader;
+	}
+
+	JPReferenceQueue* getReferenceQueue()
+	{
+		return m_ReferenceQueue;
+	}
+
+	bool getConvertStrings() const
+	{
+		return m_ConvertStrings;
+	}
+
+	// Java type resources
+	JPPrimitiveType* _void;
+	JPPrimitiveType* _boolean;
+	JPPrimitiveType* _byte;
+	JPPrimitiveType* _char;
+	JPPrimitiveType* _short;
+	JPPrimitiveType* _int;
+	JPPrimitiveType* _long;
+	JPPrimitiveType* _float;
+	JPPrimitiveType* _double;
+
+	JPClass* _java_lang_Void;
+	JPClass* _java_lang_Boolean;
+	JPClass* _java_lang_Byte;
+	JPClass* _java_lang_Character;
+	JPClass* _java_lang_Short;
+	JPClass* _java_lang_Integer;
+	JPClass* _java_lang_Long;
+	JPClass* _java_lang_Float;
+	JPClass* _java_lang_Double;
+
+	JPClass* _java_lang_Object;
+	JPClass* _java_lang_Class;
+	JPClass* _java_lang_reflect_Field;
+	JPClass* _java_lang_reflect_Method;
+	JPClass* _java_lang_Throwable;
+	JPStringType* _java_lang_String;
+
+	JPClassRef _java_lang_RuntimeException;
+	JPClassRef _java_lang_NoSuchMethodError;
+
+	jmethodID m_BooleanValueID;
+	jmethodID m_ByteValueID;
+	jmethodID m_CharValueID;
+	jmethodID m_ShortValueID;
+	jmethodID m_IntValueID;
+	jmethodID m_LongValueID;
+	jmethodID m_FloatValueID;
+	jmethodID m_DoubleValueID;
+
+private:
+
+	void loadEntryPoints(const string& path);
+	void createJVM(void* arg);	// JVM
+
+	jint(JNICALL * CreateJVM_Method)(JavaVM **pvm, void **penv, void *args);
+	jint(JNICALL * GetCreatedJVMs_Method)(JavaVM **pvm, jsize size, jsize * nVms);
+
+private:
+	friend class JPJavaFrame;
+	friend class JPClass;
+	JPContext(const JPContext& orig);
+
+	JavaVM *m_JavaVM;
+
+	// Java half
+	JPObjectRef m_JavaContext;
+
+	// Services
+	JPTypeFactory *m_TypeFactory;
+	JPTypeManager *m_TypeManager;
+	JPClassLoader *m_ClassLoader;
+	JPReferenceQueue *m_ReferenceQueue;
+	JPProxyFactory *m_ProxyFactory;
+
+public:
+	// Java Functions
+	jmethodID m_Object_ToStringID;
+	jmethodID m_Object_EqualsID;
+	jmethodID m_Object_HashCodeID;
+	jmethodID m_ShutdownMethodID;
+	jmethodID m_CallMethodID;
+	jmethodID m_Class_GetNameID;
+	jmethodID m_Context_collectRectangularID;
+	jmethodID m_String_ToCharArrayID;
+
+private:
+	bool m_IsShutdown;
+	bool m_IsInitialized;
+	bool m_ConvertStrings;
+} ;
+
+template<class jref>
+JPRef<jref>::JPRef(const JPRef& other)
+{
+	m_Context = other.m_Context;
+	if (m_Context != NULL)
+	{
+		JPJavaFrame frame(m_Context, m_Context->getEnv());
+		m_Ref = (jref) frame.NewGlobalRef((jobject) other.m_Ref);
+	} else
+	{
+		printf("NULL context in JPRef()");
```

Comment:
Shouldn't that raise? I assume this should never happen

Reply:
Agree I will change it to 

```
    JP_RAISE(PyExc_SystemError, "NULL context in JPRef()");
```

Unfortunately, I do not think I can actually test that code without instrumentation.

Reply:
We can test it later, when the instrumentation thing is in place.

Reply:
This one is harder than I expected.   The reason for the print was that I can't throw here because throw requires JPException but JPException requires JPRef creating a perfect circular dependency.   I will figure out how to break it.

Reply:
Finally got it so resolved.  The solution made unfortunately was to make a function just to raise an exception.  :sob: